### PR TITLE
Keep track of source name on all allocated SymInts

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90528
* #90381
* __->__ #90295

Wow, I had to sweat so much to get this PR out lol.

This PR enforces the invariant that whenever we allocate SymInts as part of fakeification, the SymInt is associated with a Source, and in fact we store the string source name on SymbolWithSourceName. We use 'sname' as the shorthand for source name, as 'name' is already used by sympy to name symbols.

In order to store source names, we have to plumb source names from Dynamo to PyTorch. This made doing this PR a bit bone crushing, because there are many points in the Dynamo codebase where we are improperly converting intermediate tensors into fake tensors, where there is no source (and there cannot be, because it's a frickin' intermediate tensor). I've fixed all of the really awful cases in earlier PRs in the stack. This PR is just plumbing in source names from places where we do have it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
I'm not really sure I want to plumb the snames all the way anymore. But half of this PR is useful, which is making sure we don't call fakeify when we don't have a source in hand.

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -748,31 +751,24 @@ def wrap_fake_exception(fn):
...
-            )
-        )
-    else:
-        return e
-
-
-def wrap_to_fake_tensor_and_record(e, tx, ignore_subclass=False):
-    # The not fake tensor check here is annoying - ideally, fake tensors never call this during wrapping.
-    # However, get_fake_value takes args and passes them through this, which may include fake tensors.
-    # see tree_map(fake_wrapper, args) in get_fake_value.
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:765]: why

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -748,31 +751,24 @@ def wrap_fake_exception(fn):
...
             static_shapes = True
         return wrap_fake_exception(
             lambda: make_fake_tensor(
-                e, tx.fake_mode, static_shapes, tx, ignore_subclass=ignore_subclass
+                e,
+                tx.fake_mode,
+                static_shapes,
+                tx,
+                ignore_subclass=ignore_subclass,
+                sname=sname,
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:771]: Nit: this is getting large enough to warrant its own struct

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -1053,7 +1049,11 @@ def get_fake_value(node, tx):
     from .exc import TorchRuntimeError, unimplemented, Unsupported
 
     op = node.op
-    fake_wrapper = functools.partial(wrap_to_fake_tensor_and_record, tx=tx)
+
+    def fake_wrapper(e):
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:1053]: this isn't a fake_wrapper anymore, imo worth rename

File: torch/_dynamo/variables/builder.py [hunk 4]
```diff
@@ -744,29 +751,34 @@ def _clone_input(value):
...
+                proxy.tracer.real_value_cache[proxy.node] = _clone_input(example_value)
+            # NB: If we're ignoring subclass, then the expectation is you will
+            # take the returned TensorVariable and wrap it into a more
+            # accurate TensorVariable that is able to track subclass-ness;
+            # otherwise this is wrong!
+            kwargs = {"ignore_subclass": ignore_subclass}
+            assert "source" in options
+            if options["source"] is None:
+                kwargs["static_shapes"] = True
+                kwargs["sname"] = "__constant_illegal_sname"
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:776]: is this ever checked anywhere? Why do we need a `__constant_illegal_sname` sname?

File: torch/_subclasses/meta_utils.py [hunk 5]
```diff
@@ -185,7 +185,17 @@ def meta_storage(self, s, callback):
...
+    # NB: name here is used in a conventional way by Dynamo; it corresponds
+    # precisely to the Source.name() of the tensor we're fakeifying and
+    # corresponds to a valid Python expression.  When we construct sub-names
+    # as part of this process, we will maintain this invariant!  (Even though
+    # other users of this may not need it this property to be upheld.)
+    def meta_tensor(
+        self, t, shape_env=None, callback=lambda t: t(), sname: Optional[str] = None
+    ):
+        if sname is None:
+            sname = f"__unknown_tensor{len(self.tensor_memo)}"
```
Comment (voznesenskym) [inline torch/_subclasses/meta_utils.py:197]: should we have a prefix pattern so we know where the unknowns are coming from?

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -504,23 +514,28 @@ def create_symbolic_sizes_strides_storage_offset(self, ex: torch.Tensor):
...
             # here in the smallest unbound stride case, but it's not
             # a big deal because the non-0/1 symbol immediately
             # evaporates from its duck-sizing simplification
-            s = self.create_symbol(val, simplify=False)
-            assert stride_expr is not None
-            assert isinstance(s, sympy.Symbol)
-            self.replacements[s] = stride_expr
+            s: sympy.Symbol
+            if isinstance(stride_expr, SymbolWithSourceName):
+                s = stride_expr
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:530]: is this used anywhere? did you mean for `self.replacements[s] = stride_expr` to be tabbed over?

File: torch/fx/experimental/symbolic_shapes.py [hunk 7]
```diff
@@ -530,15 +545,20 @@ def create_symintnode(self, sym: "sympy.Expr"):
     # This is guaranteed to return a symbol or its negation is a sympy.Symbol,
     # but there may be a replacement that allows it to be immediately
     # simplified
-    def create_symbol(self, val: int, *, simplify: bool = True) -> "sympy.Expr":
+    def create_symbol(self, val: int, *, simplify: bool = True, sname: str) -> "sympy.Expr":
+        assert isinstance(sname, str), f"{type(sname)} {sname}"
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:549]: nit: type is enough probably, if `sname` is a tensor this print is gonna be ugly

File: torch/fx/experimental/symbolic_shapes.py [hunk 8]
```diff
@@ -547,20 +567,37 @@ def create_symbol(self, val: int, *, simplify: bool = True) -> "sympy.Expr":
...
+        # the same symint
+        self.replacements[symbol] = self.duck_int(val)
 
         # Return the *symbol*; you're expected to apply the replacement to get
         # the simplified variable
         return symbol
 
+    # Given a concrete integer value, return the duck sized symbol associated
+    # with it; e.g., suppose we already have a tensor of size 3 in scope,
+    # which was assigned s3, then shape_env.duck_int(3) we will get back s3.
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:587]: hyper-ultra-nit: change one of the digits in the ex so folks dont think s3 is 3, s2 is 2, etc.

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (voznesenskym) [reply torch/fx/experimental/symbolic_shapes.py:530]: oh for             `sym_stride.append(self.create_symintnode(s))`

Tad hard to follow, maybe give the name a little length?

**Review by voznesenskym — APPROVED** [review]
